docs/dev/internals/bc-promise.md

```diff
@@ -0,0 +1,56 @@
+---
+title: "BC Promise"
+description: "Our backwards compatibility promise."
+---
+
+Everyone involved in the development of Contao knows about the importance of stable software and thus stable APIs.
+We try our very best to ensure that Contao updates are as smooth as they can be.
+
+Being a project that is built on top of the Symfony full-stack framework, we generally make the same
+[Backwards Compatibility Promise][SF_BC_Promise] as Symfony itself.
+We also follow the principles of [Semantic Versioning][Semver], which means that breaking changes must only be
+expected when switching to a new major version.
+
+However, in contrast to Symfony, Contao is not "just" a framework. We do not only build the tools for other
+developers but we also build tools ourselves, which are then shipped with the core distribution of Contao.
+Therefore, our BC promise deviates from the Symfony BC promise in some regards:
+
+* Our BC promise does not cover classes and methods marked as `@internal`. In most cases this concerns constructors
+  of services Contao provides. If you want to change the behaviour of a service, do not replace it with your own
+  instance of the class but instead decorate the original service (see the tip about "Composition over Inheritance").
```

Comment:
Can we somehow link to https://symfony.com/doc/current/service_container/service_decoration.html here?